tests/integration/coordinator/plugins/clickhouse/test_steps.py

```diff
@@ -99,3 +101,77 @@ async def test_retrieve_tables(ports: Ports, clickhouse_command: ClickHouseComma
                     dependencies=[],
                 ),
             ]
+
+
+class KeeperMapInfo(NamedTuple):
+    context: StepsContext
+    clickhouse_client: ClickHouseClient
+    user_client: ClickHouseClient
+
+
+@pytest.fixture(name="keeper_table_context")
+async def fixture_keeper_table_context(
+    ports: Ports, clickhouse_command: ClickHouseCommand, minio_bucket: MinioBucket
+) -> AsyncIterator[KeeperMapInfo]:
+    async with (
+        create_zookeeper(ports) as zookeeper,
+        create_clickhouse_cluster(zookeeper, minio_bucket, ports, ["s1"], clickhouse_command) as clickhouse_cluster,
+    ):
+        clickhouse = clickhouse_cluster.services[0]
+        admin_client = get_clickhouse_client(clickhouse)
+        await setup_cluster_users([admin_client])
+        for statement in [
+            b"CREATE DATABASE `keeperdata` ENGINE = Replicated('/clickhouse/databases/keeperdata', '{my_shard}', '{my_replica}')",
+            b"CREATE TABLE `keeperdata`.`keepertable` (thekey UInt32, thevalue UInt32) ENGINE = KeeperMap('test', 1000) PRIMARY KEY thekey",
+            b"INSERT INTO `keeperdata`.`keepertable` SELECT *, materialize(1) FROM numbers(3)",
+            b"CREATE USER bob IDENTIFIED WITH sha256_password BY 'secret'",
+            b"GRANT INSERT, SELECT, UPDATE, DELETE ON `keeperdata`.`keepertable` TO `bob`",
+        ]:
+            await admin_client.execute(statement)
+        user_client = HttpClickHouseClient(
+            host=clickhouse.host,
+            port=clickhouse.port,
+            username="bob",
+            password="secret",
+            timeout=10,
+        )
+        step = RetrieveDatabasesAndTablesStep(clients=[admin_client])
+        context = StepsContext()
+        databases_tables_result = await step.run_step(Cluster(nodes=[]), context=context)
+        context.set_result(RetrieveDatabasesAndTablesStep, databases_tables_result)
+        yield KeeperMapInfo(context, admin_client, user_client)
+
+
+async def test_keeper_map_table_select_only_setting_modified(keeper_table_context: KeeperMapInfo) -> None:
+    steps_context, admin_client, user_client = keeper_table_context
+    read_only_step = KeeperMapTablesReadOnlyStep(clients=[admin_client], allow_writes=False)
+    # After the read-only step, the user should only be able to select from the table
+    await read_only_step.run_step(Cluster(nodes=[]), steps_context)
+    with pytest.raises(ClickHouseClientQueryError, match=".*ACCESS_DENIED.*"):
+        await user_client.execute(
+            b"INSERT INTO `keeperdata`.`keepertable` SETTINGS wait_for_async_insert=1  SELECT *, materialize(2) FROM numbers(3)"
+        )
+    with pytest.raises(ClickHouseClientQueryError, match=".*ACCESS_DENIED.*"):
```

Comment:
So, we are basically do it for the "ClickHouse connector for Kafka". I have not read it's code, but I suspect some new problems in the normal operation.
What does exactly once mean: every part is processed, every part is not processed 2 times.
Before this commit if there are not backup restore and errors during the operation it was exactly once.
Now, during backup some part will not be able to say that the work is done.
How exactly will it behave?
Because it theoretically could
1. fail instantly and someone else repeats the whole processing
2. be restarted and the processing is done again
If so, it is no better than it was, because previously it  was not happening during normal operation

Reply:
The sink connector pauses writes to ClickHouse for these exceptions: https://github.com/ClickHouse/clickhouse-kafka-connect/blob/main/src/main/java/com/clickhouse/kafka/connect/util/Utils.java#L68

If the exception is retriable, the connector won't commit the offsets to the KeeperMap table, and won't try to write that part to ClickHouse until later. The design docs describes the state machine transitions: https://github.com/ClickHouse/clickhouse-kafka-connect/blob/main/docs/DESIGN.md

Reply:
Critically, the `ACCESS_DENIED` raises exception code 497 which isn't included in the retriable exceptions list. So we need to change the approach a bit.